src/zarr/core/metadata/converter/converter_v2_v3.py

```diff
@@ -0,0 +1,256 @@
+    ZMETADATA_V2_JSON,
+    ZarrFormat,
+)
+from zarr.core.dtype.common import HasEndianness
+from zarr.core.group import Group, GroupMetadata
+from zarr.core.metadata.v2 import ArrayV2Metadata
+from zarr.core.metadata.v3 import ArrayV3Metadata
+from zarr.core.sync import sync
+from zarr.registry import get_codec_class
+from zarr.storage import StoreLike
+from zarr.storage._common import make_store_path
+
+logger = logging.getLogger(__name__)
+
+
+def convert_v2_to_v3(
+    store: StoreLike,
+    path: str | None = None,
+    storage_options: dict[str, Any] | None = None,
+    dry_run: bool = False,
+) -> None:
+    """Convert all v2 metadata in a zarr hierarchy to v3. This will create a zarr.json file at each level
+    (for every group / array). V2 files (.zarray, .zattrs etc.) will be left as-is.
+
+    Parameters
+    ----------
+    store : StoreLike
+        Store or path to directory in file system or name of zip file.
+    path : str | None, optional
+        The path within the store to open, by default None
+    storage_options : dict | None, optional
+        If the store is backed by an fsspec-based implementation, then this dict will be passed to
+        the Store constructor for that implementation. Ignored otherwise.
+    dry_run : bool, optional
+        Enable a 'dry run' - files that would be created are logged, but no files are actually created.
+    """
+
+    zarr_v2 = zarr.open(store=store, mode="r+", path=path, storage_options=storage_options)
+    convert_array_or_group(zarr_v2, dry_run=dry_run)
+
+
+def convert_array_or_group(zarr_v2: Array | Group, dry_run: bool = False) -> None:
+    """Convert all v2 metadata in a zarr array/group to v3. Note - if a group is provided, then
+    all arrays / groups within this group will also be converted. A zarr.json file will be created
+    at each level, with any V2 files (.zarray, .zattrs etc.) left as-is.
+
+    Parameters
+    ----------
+    zarr_v2 : Array | Group
+        An array or group with zarr_format = 2
+    dry_run : bool, optional
+        Enable a 'dry run' - files that would be created are logged, but no files are actually created.
+    """
+    if not zarr_v2.metadata.zarr_format == 2:
+        raise TypeError("Only arrays / groups with zarr v2 metadata can be converted")
+
+    if isinstance(zarr_v2.metadata, GroupMetadata):
+        # process members of the group
+        for key in zarr_v2:
+            convert_array_or_group(zarr_v2[key], dry_run=dry_run)
+
+        # write group's converted metadata
+        group_metadata_v3 = GroupMetadata(
+            attributes=zarr_v2.metadata.attributes, zarr_format=3, consolidated_metadata=None
+        )
+        sync(_save_v3_metadata(zarr_v2, group_metadata_v3, dry_run=dry_run))
+
+    else:
+        # write array's converted metadata
+        array_metadata_v3 = _convert_array_metadata(zarr_v2.metadata)
+        sync(_save_v3_metadata(zarr_v2, array_metadata_v3, dry_run=dry_run))
+
+
+async def remove_metadata(
```

Comment:
Am I right in thinking this isn't required by the conversion functionality? If not, I'd be tempted not to implement it at least in this PR to keep this PR tightly scoped and therefore easier to review. I have some comments/questions about performance implications here, but thought maybe that could be discussed in a follow up PR?

Reply:
Sorry, I just read the comment thread, feel free to ignore this and leave clear here!